---
title: "BARBI Workflow"
author:
  name: Pratheepa Jeganathan
  affiliation: Department of Statistics, Stanford University
  email: jpratheepa31@gmail.com
package: BARBI
abstract: >
  BAyesian Reduction in Background Interference (BARBI) provides a reproducible and automated process to remove contaminating DNA from metagenomic shotgun sequencing data in the setting of ultra low-biomass samples (After removing Eukaryotic pathogens).
output:
  BiocStyle::html_document:
    toc_float: true
vignette: >
    %\VignetteIndexEntry{BARBI Workflow}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
bibliography: BARBI.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

Install [R](https://www.r-project.org/) and [RStudio](https://www.rstudio.com/). Open this `Rmd` `r Githubpkg("rstudio/rmarkdown")` file in RStudio. Then run the following code to install all required packages.

```{r install_packages,eval=FALSE}
pkgs <- c("DESeq2","phyloseq","dplyr",
          "tidyr","R.utils","BiocParallel",
          "doParallel","parallel","HDInterval",
          "grid","xtable","gtable",
          "gridExtra","BiocStyle","magrittr")
source("http://bioconductor.org/biocLite.R")
biocLite(setdiff(pkgs,installed.packages()))
devtools::install_github("PratheepaJ/BARBI")
```

Load packages:
```{r load_packages}
library(BARBI)
library(devtools)
library(phyloseq)
library(DESeq2)
library(dplyr)
library(tidyr)
library(R.utils)
library(BiocParallel)
library(doParallel)
library(parallel)
library(HDInterval)
library(grid)
library(xtable)
library(gtable)
library(gridExtra)
library(BiocStyle)
library(magrittr)

#devtools::load_all(".")
```

#      Load the real data stored as a phyloseq object in the ``BARBI`` package or use your own data stored as a phyloseq object

##      Load the phyloseq 

Analysis of shotgun-metagenomic sequencing samples that is a subset of SIRS patients from @cheng2017integration.

```{r read-phylo}
ps <- psSub
```

We remove library controls. 

```{r remove_unwanted_controls}
ps <- subset_samples(ps,Sample_Type=!"Library_Control")
```

We identified three different blocks/batches of processing samples using a PCA on ranks of abundance (though we expect a user to know the blocks/batches in advance for each samples). 

```{r adding_blocks}
set1 <- c("1","2","3","4","11","12")
set2 <- c("5","6","7","8","9","10")
setP <- "P"
ext.num <- sample_data(ps)$Extraction_Number

blocks <-ifelse(ext.num %in% set1, "Set1", ifelse(ext.num %in% set2, "Set2", "SetP"))

sample_data(ps)$block <- blocks
```

We identified the species not in any single plasma sample and removed them from the phyloseq object and label them as contaminants.

```{r filter_taxa}
ps <- prune_taxa(taxa_sums(ps)>0, ps)
ps_plasma <- subset_samples(ps,Sample_Type%in%"Plasma")
prevTaxaP <- apply(otu_table(ps_plasma), 1 ,function(x){sum(x>0)})

Contaminants1 <- names(prevTaxaP)[prevTaxaP==0]
ps <- prune_taxa(prevTaxaP>0,ps)
```

We set that the samples are on the columns and species are on the rows of `otu_table`. Then, we checked the distribution of library size/sample depth to see whether we can drop any sample with less library size.

```{r filter-samples}
if(dim(otu_table(ps))[1]!=ntaxa(ps)){otu_table(ps) <- t(otu_table(ps))}
totalReads <- colSums(otu_table(ps))
hist(log(totalReads),yaxs="i",xaxs="i",main="Distribution of total reads per sample",breaks=50)
```

We make sure that we filter only the SIRS/plasma, negative control, and healthy control samples.

```{r choose_samples}
ps <- subset_samples(ps,Sample_Type %in% c("Plasma","Control"))
```

Summary of plasma and negative control samples in each block.

```{r summary_stat}
table(sample_data(ps)$Sample_Type,sample_data(ps)$block)
```

#       Bayesian Inference 

##      Prepare the phyloseq object for the Bayesian inference

We use the Bayesian inference to identify contaminants in each block because DNA contaminants are unique to each batch. 

Thus, we split the phyloseq object corresponding to each block and store as a list of phyloseq objects, `psByBlock`. 

We select negative control samples from each block and store as a list of phyloseq objects, `psNCbyBlock`. 

We select all species that has prevalence of zero in all negative control samples for each block and store as a list of phyloseq objects, `psallzeroInNC`.

We select all plasma samples from each block and store as a list of phyloseq objects, `psPlByBlock`.

```{r list_of_phyloseq}
psBlockResult <- psBlockResults(ps, sampleTypeVar = "Sample_Type",sampleName = "SampleCode",blockVar = "block")
psByBlock <- psBlockResult[[1]]
psNCbyBlock <- psBlockResult[[2]]
psallzeroInNC <- psBlockResult[[3]]
psPlByBlock <- psBlockResult[[4]]

#       subset for the workflow
psPlByBlock <- lapply(psPlByBlock, function(x){
        subsetSample <- as.character(sample_data(x)$SampleCode[c(1,5)])
        x <- prune_samples(subsetSample,x)
        return(x)
})
```

##      Estimate the distribution parameters of the DNA contamination in negative control samples

We estimate the gamma distribution parameters for the intensity of contaminants usning the negative control samples for each block. 

`add1 == TRUE` corresponds to adding 1 to the otu table and `add1 == TRUE` corresponds to choosing 'poscounts' for saize factor calculation method in `DESeq`.

```{r estimate_Cont_Intensity_ncontrols}
alphaBetaNegControl <- alphaBetaNegControl(psNCbyBlock = psNCbyBlock, add1 = TRUE)
```

##      Estimate the distribution parameters for the intensity of contamination in a plasma sample

For each plasma sample, we estimate the gamma distribution parameters for the intensity of contamination using the scaling property of the gamma distribution.

`stringent == TRUE` corresponds to `add1 == TRUE` and use the gamma parameters as in the negative control samples.

```{r estimate_Cont_Intensity_plasma}
num_blks <- length(alphaBetaNegControl)
blks <- seq(1, num_blks) %>% as.list

gammaPrior_all_blks <- lapply(blks, function(x){
        gammaPrior <- alphaBetaContInPlasma(psPlByBlock = psPlByBlock, psallzeroInNC = psallzeroInNC, blk = x, alphaBetaNegControl=alphaBetaNegControl, stringent = TRUE)
        return(gammaPrior)
})
```

##      Sampling from the posterior for the intensity of true signal

For all samples and for all taxa, we sampling from the posterior for the intensity of true signal using the Metropolis-Hasting MCMC. We can choose the number of MCMC using the option `itera`.

We save the gamma prior for the intensity of contamination and the posterior samples.

```{r sampling_post_true_int,eval=FALSE}
t1 <- proc.time()

post_all_blocks <- lapply(blks,function(x){
        post_int_all_taxa <- samplingPosterior(
                psPlByBlock = psPlByBlock,
                blk = x,
                gammaPrior_Cont = gammaPrior_all_blks[[x]],
                itera = 100)
        return(post_int_all_taxa)
})

proc.time()-t1

gammaPrior_posTrueSing_all_blocks <- list(gammaPrior_all_blks,post_all_blocks)

saveRDS(gammaPrior_posTrueSing_all_blocks, file= "./gammaPrior_posTrueSing_all_blocks.rds")
```

#       Display the results

##     Make summaires from the BARBI results.

We can choose the number of MCMC to be removed using the option `burnIn` and it must be less than `itera`.

We can choose the coverage probability to construct the highest posterior density interval using the option `cov.pro`.

We can choose the option `Use_SNR` to tell whether the filtering of species is also based on the signal-to-noise ratio, otherwise, the filtering is based on the lower limit of the true signal and the upper limit of the contaminant. 

```{r make_tables, eval=FALSE}
itera <- 100
burnIn <- 10
cov.pro <- .95
Use_SNR <- FALSE


gammaPrior_posTrueSing_all_blocks <- readRDS("./gammaPrior_posTrueSing_all_blocks.rds")

gammaPrior_all_blks <- gammaPrior_posTrueSing_all_blocks[[1]]
post_all_blocks <- gammaPrior_posTrueSing_all_blocks[[2]]

finalPosTaxa <- data.frame()
final_results_list <- list()

for(blk in 1:num_blks){
        
        taxa_post_all_sam <- post_all_blocks[[blk]]
        gammPrior <- gammaPrior_all_blks[[blk]]
    
        total_summary_table <- NULL 
        
        all_real_taxa <- character()

        for(sam in 1:nsamples(psPlByBlock[[blk]])){
          
                taxa_post <- taxa_post_all_sam[[sam]]
                acceptance <- list()
                exp_post_s <- list()
                lower.s <- list()
                upper.s <- list()
                upper.b <- list()
                lower.b <- list()
                sd.s <- list()
                naive.exp <- list()
                all.zero.nc <- list()
                  for(taxa in 1:length(taxa_post)){
                          burnIn  <- burnIn
                          acceptance[[taxa]]  <-  1-mean(duplicated(taxa_post[[taxa]][-(1:burnIn),]))
                  
                          exp_post_s[[taxa]] <- mean(taxa_post[[taxa]][-(1:burnIn),])
                            
                          hdi.v <- hdi(taxa_post[[taxa]][-(1:burnIn),], credMass = cov.pro)
                          lower.s[[taxa]] <- round(hdi.v[1], digits = 0)
                          upper.s[[taxa]] <- round(hdi.v[2], digits = 0)
                          
                          hdi.b <- hdi(rgamma((itera-burnIn+1), shape=gammPrior[[sam]][[1]][taxa], rate = gammPrior[[sam]][[2]][taxa]), credMass = cov.pro)
                          lower.b[[taxa]] <- round(hdi.b[1], digits = 0)
                          upper.b[[taxa]] <- round(hdi.b[2], digits = 0)
                          
                          sd.s[[taxa]] <- sd(taxa_post[[taxa]][-(1:burnIn),])
                          sd.s[[taxa]] <- round(sd.s[[taxa]], digits = 0)
                          
                          naive.exp[[taxa]] <- gammPrior[[sam]][[3]][taxa] - (gammPrior[[sam]][[1]][taxa]/gammPrior[[sam]][[2]][taxa])
                          naive.exp[[taxa]] <- round(naive.exp[[taxa]], digits = 0)
                          
                          all.zero.nc[[taxa]] <-  gammPrior[[sam]][[5]][taxa]
                  }

                mappingFileByBlock <-lapply(psByBlock, function(x){
                        as.data.frame(x@sam_data)
                        })
                sampleDepth <- as.numeric(mappingFileByBlock[[blk]][sample_names(psPlByBlock[[blk]])[sam], "Reads"])

                df <- data.frame(Species=taxa_names(psPlByBlock[[blk]]),
                                 xj=as.numeric(gammPrior[[sam]][[3]]),
                                 l.s=unlist(lower.s),
                                 u.s=unlist(upper.s),
                                 l.b=unlist(lower.b),
                                 u.b=unlist(upper.b),
                                 sd.s=unlist(sd.s),
                                 naive.exp=unlist(naive.exp),
                                 all.zero.nc=unlist(all.zero.nc),
                                 Sample_Depth = rep(sampleDepth,length(unlist(upper.s))),
                                 Sample = rep(sample_names(psPlByBlock[[blk]])[sam],
                                              length(unlist(upper.s))))
  
                  df <- arrange(filter(df,(l.s>u.b)&(l.s>0)), desc(xj))
                  
                  # Filter Out Viral Taxa
                  ps.tax.df <- as.data.frame(ps@tax_table)
                  ps.tax.df.viruses <- filter(ps.tax.df, 
                                              Superkingdom=="k__Viruses")
                  list.of.viruses <- as.character(ps.tax.df.viruses[,"Species"])
                  df <- arrange(filter(df, 
                                       !(Species %in% list.of.viruses)),
                                desc(xj))
                  
                  
                  finalPosTaxa <- rbind(finalPosTaxa, df)
                  
                  
                  if(dim(df)[1]==0){
                          df <- data.frame(Species="Negative",
                                           xj="Negative",
                                           l.s="Negative",
                                           u.s="Negative",
                                           l.b ="Negative", u.b="Negative", sd.s = "Negative", naive.exp = "Negative",
                                           all.zero.nc = "Negative",
                                           Sample_Depth=sampleDepth,
                                           Sample = sample_names(psPlByBlock[[blk]])[sam])
                  }                  
                  
                  all_real_taxa <- c(all_real_taxa, 
                                     as.character(df$Species))

                  #create a line.summary variable that summarizes df for each sample in 1 line
                  sam.summary <- character()
                  for (i in 1:nrow(df)) {
                    
                    if (df[i, "Species", drop = FALSE]=="Negative") {
                      sam.summary <- "; Negative"
                    }
                    else {
                    taxa.summary <- paste(df[i, "Species"], 
                                          " (", df[i, "xj"], 
                                          ", ", df[i, "l.s"], 
                                          ", ", df[i, "u.b"], 
                                          ", ", df[i, "all.zero.nc"], 
                                          ")",  sep="")
                    sam.summary <- paste(sam.summary, "; ", 
                                         taxa.summary, sep = "")
                    }
                  }
                  
                  
                  sam.summary <- substring(sam.summary, 3)
                  total_summary_table <- rbind(total_summary_table,       data.frame(sample_names(psPlByBlock[[blk]])[sam],sam.summary))                 
                  filname <- paste("./",
                                   sample_names(psPlByBlock[[blk]])[sam],
                                   ".png",
                                   sep="")
                 
                  png(filname, 
                      height = 600, 
                      width = 750)
                  
                 df.mod <- df
                 df.mod <- subset(df.mod, 
                                  select = c(Species, 
                                             xj, l.s, u.b, all.zero.nc))
                 
                colnames(df.mod) <- c("Species", 
                                      "Raw Reads", 
                                      "Lower Limit Real",
                                      "Upper Limit Contaminant", 
                                      "In Neg. Cont.")
                 
                df.p <- tableGrob(df.mod)
                
                title <- textGrob(sample_names(psPlByBlock[[blk]])[sam],
                                  gp = gpar(fontsize = 14))
                
                padding <- unit(0.5,"line")
                df.p <- gtable_add_rows(df.p,
                                        heights = grobHeight(title) + padding,
                                        pos = 0)
                  df.p <- gtable_add_grob(
                          df.p, 
                          list(title),
                          t = 1, l = 1, 
                          r = ncol(df.p))
                  
                  grid.newpage()
                  grid.draw(df.p)
                  dev.off()
                  
                  
  }
        
        all_real_taxa <- unique(all_real_taxa)
        final_results_list[[blk]] <- list(total_summary_table, all_real_taxa)
}

total_summary_table_list  <- list() 
total_summary_table_list  <- lapply(final_results_list, 
                                    function(x) {
                                            x[[1]]
                                            })

total_summary_table_all <- do.call("rbind", 
                                   total_summary_table_list )

row.names(total_summary_table_all) <- total_summary_table_all[,1]
colnames(total_summary_table_all) <- c("Sample", 
                                       "Species (Raw Reads, Lower Bound Background Reads Estimate, Upper Bound Real Reads Estimate, Presence in Neg. Controls)")
total_summary_table_all <- total_summary_table_all[order(row.names(total_summary_table_all)), , drop = FALSE]

write.table(total_summary_table_all, 
            paste("./total_summary_table_95", ".txt", sep=""), 
            sep="\t", 
            col.names=TRUE, 
            row.names = FALSE)

saveRDS(finalPosTaxa,
        file="./finalPosTaxa.rds")
```

# Session Info {.unnumbered}

```{r session_info}
sessionInfo()
```
